Implement class to create scenario from survey answers

paulapreuss opened the pull request:
This is WIP to implement a class that will take the answers from the survey and create the necessary scenario `.csv` from the data. I restructured and added some things to the code based on what was already existing / what I assume was also similar to what you had in mind @Bachibouzouk.

# The premise
Currently, the class assumes that there will be a mapping dictionary such as 
```
SURVEY_ANSWER_COMPONENT_MAPPING = {
    # "question_id": {"yes": "component_name"},
    # "question_id": {"no": "component_name", "yes": "other_component_name"},
    "1": {
        "photovoltaic_system": "pv_ground_mounted",
        "battery_system": "battery-storage",
    },
    "1.1": {"pv_ground_mounted/capacity": TYPE_FLOAT},
    "1.2": {"battery_storage/capacity": TYPE_FLOAT},
```
where the **first level** relates to a question to **add components** and the **sub-questions** relate to the **components attribute**, which are described in a **{component}/{attribute} format**. It also assumes that all questions will be numbered (e.g. 3.1a and 3.1b are not good question numbers). These rules are important for the current logic to make sense.
Furthermore, the format for the input of the class is currently the output of `form.cleaned_data` from the survey input, which is currently a dictionary with the format `{"criteria_*": answer, ...}`.

# The process
The idea is that the survey answers are used to build the `.csv` files for `oemof-tabular`, which will then later be converted to a datapackage. The components are based on the existing components in the `component_library`. I added a few more just for testing puposes. Once the survey is filled out, the process (so far) goes as follows:
1. `process_survey` - The survey is processed to create a dictionary that contains the components to add as keys and nested dictionaries containing the attributes to be changed
2. `add_components` - The previously created dictionary is used to create files containing the corresponding components. This is based on code that you wrote. Then the given attributes are changed based on the values from the surveys. 
3. `add_buses` - The component files are scanned to find to which buses the components are connected to. It then creates a bus file containing all of these buses. It may also be necessary to add excess sinks based on the buses but this is not implemented yet.
4. `add_sequences` - The component files are scanned to see which files contain a `profile` column. It then creates sequences files containing the profiles for the corresponding components. Currently, dummy timeseries with random values are generated in `fetch_component_timeseries`. In the future, the timeseries should be fetched from the database based on the component (weather data, uploaded load data etc.).

Once the code looks good, I will migrate it to `WEFEgui` to properly connect it to the GUI and handle the database-related work over there.

# The output
To test the current status, run `python build_scenario.py` from `component_library\scripts`. You can play around with the `test_survey` contained in the file. The code should produce a folder `some_scenario_name` (placeholder) within the directory containing the created `oemof-tabular` files in the correct format.

paulapreuss committed:
Create `ScenarioBuilder` class to process survey answers

paulapreuss committed:
Migrate existing methods to `ScenarioBuilder` class

paulapreuss committed:
Edit component attributes while adding components

paulapreuss committed:
Adapt `add_components` to new data flow

paulapreuss committed:
Add function to update attributes

paulapreuss committed:
Add buses based on added components

paulapreuss committed:
Add code to test script

paulapreuss committed:
Add sequences files based on system components

Bachibouzouk committed:
Add dummy components to test mapping

Bachibouzouk committed:
Set components which still have comments aside

Bachibouzouk committed:
Save survey answers from app

Bachibouzouk committed:
Load survey answers from dumped forms

Bachibouzouk committed:
Display errors by datapackege errors

Bachibouzouk commented on component_library/scripts/build_scenario.py:
```diff
@@ -1,85 +1,326 @@
+                file = os.path.join(scenario_component_folder, filename)
+                components = pd.read_csv(file)
+                # Look for columns pertaining to bus connections
+                bus_cols = components.filter(regex="^(bus|from_bus_.*|to_bus_.*)$").columns
```
here the carrier would also be good to get as information

Bachibouzouk committed:
Add dummy survey answer

Bachibouzouk committed:
Fix component names for dummy example

Bachibouzouk reviewed (approved):
Thanks @paulapreuss for this stepping stone! 

I rebased on the latest merges and added a few commits to have the fonctionnality working.

At this stage we need to start generate various usecases using this and refine the functions as the usecases come. I suggest we do so in subsequent PRs

JulianBarinton committed:
start survey review and improvement

JulianBarinton committed:
minor revisions of survey, WIP

JulianBarinton committed:
survey reforming water pump part (WIP)

JulianBarinton committed:
WIP water pump adaption in survey, current error: exception value 3.4

JulianBarinton committed:
WIP CROP and Irrigation Type template

JulianBarinton committed:
add technical elements of final Aiwa case study to component library

JulianBarinton committed:
update component library with Aiwa specific components (electricity grid, bottled water, condensed water)

JulianBarinton committed:
WIP SURVEY_ANSWER_COMPONENT_MAPPING, currently at 4

JulianBarinton committed:
WIP SURVEY_ANSWER_COMPONENT_MAPPING currently 7.3

JulianBarinton committed:
finalize first draft of component mapping

Bachibouzouk commented on app/survey/survey_copy.py:
```diff
@@ -0,0 +1,686 @@
+import copy
```
you copied the survey.py file ? The point of version control is that you don't need to do this :)

Bachibouzouk commented on app/survey/survey.py:
```diff
@@ -359,7 +451,7 @@ def generate_water_questions(suffixes, survey_questions):
                  "Its primary function is to support specific activities without being directly consumed by humans.",
         "question_id": "2",
-        "possible_answers": ["yes", "no"],
+        "possible_answers": ["Yes", "No"],
```
If you change "yes" to "Yes" here and not in the subquestion mapping below then the link between question's answer and subquestion is broken :) @JulianBarinton

Bachibouzouk commented on app/survey/survey.py:
```diff
@@ -140,89 +229,101 @@
-            "unprotected_spring",
-            "public_tap_water",
-            "desalinated_seawater",
+            "groundwater well"
```
the `,` at the end of the line is important to separate the components @JulianBarinton :)

Bachibouzouk committed:
Fix survey formatting

Bachibouzouk committed:
Turn the subquestion function for water generic

Bachibouzouk committed:
Add an option to the update command

If answers are already provided to a scenario, then the SurveyQuestions
are not updated in the browser. When we work on the survey we should use
scenario 1 and clear all answer at each update of the questions

Bachibouzouk committed:
Add formating test of the questions

Bachibouzouk committed:
Enable generation of matrix subquestions

Bachibouzouk committed:
Fix question display type

Bachibouzouk committed:
Rename variable

Bachibouzouk committed:
Update SurveyQuestion model

Bachibouzouk committed:
Display the questions in the same order they are provided

Bachibouzouk committed:
Assign maxtrix classes for questions

Bachibouzouk committed:
Add migration

Bachibouzouk committed:
Rename SEC_DW to 5.3

Bachibouzouk committed:
WIP matrix vizualization

Bachibouzouk committed:
Link matrix supra_question with different js onchange

Bachibouzouk commented:
@paulapreuss - I pushed the latest changes as a WIP commit. You need to run `python manage.py update_survey_questions --update --dev` before `python manage.py runserver`. Then you should see the table of input fields (I called it a matrix). There are three of them because they are for answer "No" to question (one instance for water) 2 as well as "Yes" (2 instances, one for drinking water and another for service water). To test functionnality, just answer "No" to question 2 and then scroll down to question 5, the rows of the matrix should only be seen if the user tick the corresponding water treatment technology. 

One should see the rows and column headers (explicit questions 5.1, 5.2 and 5.3 for the columns headers, and the name of the technology for the rows headers.

One should see nothing by default, the matrix should start to appear as the user starts to tick answers. (need to bind js here)

I have added a call to a js function which is to be created yet: `triggerMatrixSubQuestion` (it might be inspired from `triggerSubQuestion`)

Whatever help in the form of recommendation or implementation is welcome :) I won't work on it before Thursday 5pm

paulapreuss committed:
Move javascript into separate file

paulapreuss committed:
Refactor trigger into separate functions

paulapreuss commented:
I've separated some of the content of `triggerSubQuestion` with the idea that both `getSelectedValues` and `getRelevantSubquestions` can probably be reused for `triggerMatrixSubQuestion`, and then the visibility toggling is what becomes a little more complicated

paulapreuss commented:
Another approach created while chatting with ChatGPT that I think might be worth a shot without having looked too much into it, but would have to be validated first:

## Approach: Use `formset_factory` for Dynamic Forms
- Instead of creating multiple static forms, generate forms on demand.
- Only create forms when needed (e.g., when a user selects "Yes" for water differentiation).
- Use JavaScript to trigger form creation dynamically in the frontend.

### Step 1: Modify Your Django View to Use `formset_factory`
Django's formset_factory helps generate multiple forms dynamically without hardcoding none/a/b.

```
from django.forms import formset_factory
from .forms import SurveyForm

def survey_view(request):
    WaterFormSet = formset_factory(SurveyForm, extra=0)  # Start with 0 forms

    if request.method == "POST":
        formset = WaterFormSet(request.POST)
        if formset.is_valid():
            formset.save()  # Save responses
    else:
        formset = WaterFormSet()  # No forms initially

    return render(request, "survey_template.html", {"formset": formset})
```
### Step 2: Update the Template
Instead of generating none/a/b forms in advance, use JavaScript to trigger their creation dynamically.
```
<form method="post">
    {% csrf_token %}

    <!-- Main survey question -->
    <label>Do you want to differentiate between water types?</label>
    <input type="radio" name="differentiate_water" value="yes" id="diff_yes"> Yes
    <input type="radio" name="differentiate_water" value="no" id="diff_no"> No

    <!-- Container where extra forms appear -->
    <div id="dynamic_forms_container"></div>

    <button type="submit">Submit</button>
</form>

<script>
document.querySelectorAll('input[name="differentiate_water"]').forEach((radio) => {
    radio.addEventListener("change", function() {
        let container = document.getElementById("dynamic_forms_container");
        container.innerHTML = ""; // Clear previous forms
        
        if (this.value === "yes") {
            fetch("/get_dynamic_forms/")  // Fetch new formset from Django
                .then(response => response.text())
                .then(html => container.innerHTML = html);
        }
    });
});
</script>
```

### Step 3: Create a Django View for Dynamic Forms
This view will return new forms only when needed.
```
from django.http import JsonResponse
from django.template.loader import render_to_string

def get_dynamic_forms(request):
    WaterFormSet = formset_factory(SurveyForm, extra=2)  # Generate 'A' and 'B' forms
    formset = WaterFormSet()

    html = render_to_string("dynamic_forms_partial.html", {"formset": formset})
    return JsonResponse({"html": html})
```

### Step 4: Create a Partial Template (dynamic_forms_partial.html)
This template is loaded only when the user selects "Yes".
```
{% for form in formset %}
    <div class="form-container">
        {{ form.as_p }}
    </div>
{% endfor %}
```

Bachibouzouk committed:
Add matrix structure

Bachibouzouk committed:
Add css to hide disabled questions

Bachibouzouk committed:
Fix POST save of survey

Bachibouzouk committed:
Fix subquestion cleaning

Bachibouzouk committed:
Fix matrix display

Bachibouzouk committed:
Update compute.py file

Bachibouzouk committed:
Fix question 7

Bachibouzouk committed:
Lint with black

Bachibouzouk committed:
Update matrix style

Bachibouzouk committed:
Remove decentralized waste water treatment plant as technology

Bachibouzouk committed:
Fix Matrix layout

Now the forms in a matrix row are tagged with an extra class label_row_
and this is used to display/hide one row at the time

Bachibouzouk committed:
Add irrigation question

Bachibouzouk committed:
Add variable_name in the survey to build automatically the answer mapping

Bachibouzouk committed:
Fix typos in survey questions

Bachibouzouk committed:
Add test of consistency between survey questions and answer mapping

For example all question of survey must be present in the answer mapping
and all missing or extra questions in the answer mapping must be
reported

Bachibouzouk committed:
Fix survey numbering url

When someone visit 127.0.0.1:8000 there is now an automatic scenario id
assignation and a url redirect to the correct page

Bachibouzouk committed:
Update the survey question to DB command

Ignore the attributes answer_map_to and variable_name which are only
used for the result processing

Bachibouzouk committed:
Add answer_map_to to a question

This is to know if from the answer a component must be insterted inside
the energy system

JulianBarinton committed:
adding wastewater and water component model drafts to component library; still water production techs have to be added

Bachibouzouk committed:
Start changing

Bachibouzouk committed:
Make sure the subQuestion exists

Bachibouzouk committed:
Add a fonction to change the suffix of a question

Bachibouzouk committed:
Update the survey commands

Bachibouzouk committed:
Fix matrix question initial display

Bachibouzouk committed:
Fix text inputs in the survey

Bachibouzouk committed:
Fix GET error when no survey saved

Bachibouzouk committed:
Remove unused file

Bachibouzouk committed:
Fix survey structure

Make the subquestions of 3 dependent of the type of water source

Bachibouzouk committed:
Reassign categories

Bachibouzouk committed:
Update README

MaaJoo13 commented:
Already integrated into https://github.com/rl-institut/WEFEConfigurator/pull/13
